packages/react/src/internal/useControllableState.js

```diff
@@ -8,68 +8,90 @@
+ * @param {Function} config.onChange - an optional function that is called when
+ * the value of the state changes. This is useful for communicating to parents of
+ * controlled components that the value is requesting to be changed.
+ * @param {any} config.value - a controlled value. Omitting this means that the state is
+ * uncontrolled
+ * @returns {[any, Function]}
+ */
+export function useControllableState({
+  defaultValue,
+  name = 'custom',
+  onChange,
+  value,
+}) {
+  const [state, internalSetState] = useState(value ?? defaultValue);
+  const controlled = useRef(null);
 
-  // If the owner is controlling the component prop value, keep the controlled
-  // state value and the internal state value in sync.
-  //
-  // We guard on `undefined` to prevent downstream breakage of controlled
-  // components (like <input>). When the controlled state switches to
-  // `undefined`, we are moving from controlled to uncontrolled.
-  if (
-    controlled.current === true &&
-    controlledState !== state &&
-    controlledState !== undefined
-  ) {
-    internalSetState(controlledState);
+  if (controlled.current === null) {
+    controlled.current = value !== undefined;
   }
 
   function setState(stateOrUpdater) {
-    if (controlled.current === true) {
-      controlledSetState(stateOrUpdater);
-    } else {
-      internalSetState(stateOrUpdater);
+    const value =
+      typeof stateOrUpdater === 'function'
+        ? stateOrUpdater(state)
+        : stateOrUpdater;
+
+    if (controlled.current === false) {
+      internalSetState(value);
+    }
+
+    if (onChange) {
+      onChange(value);
     }
   }
 
   useEffect(() => {
+    const controlledValue = value !== undefined;
+
     // Uncontrolled -> Controlled
     // If the component prop is uncontrolled, the prop value should be undefined
-    if (controlled.current === false && controlledState !== undefined) {
+    if (controlled.current === false && controlledValue) {
       warning(
         false,
-        'A component is changing an uncontrolled component to be controlled. ' +
+        'A component is changing an uncontrolled %s component to be controlled. ' +
           'This is likely caused by the value changing to a defined value ' +
           'from undefined. Decide between using a controlled or uncontrolled ' +
           'value for the lifetime of the component. ' +
-          'More info: https://reactjs.org/link/controlled-components'
+          'More info: https://reactjs.org/link/controlled-components',
+        name
       );
     }
 
     // Controlled -> Uncontrolled
     // If the component prop is controlled, the prop value should be defined
-    if (controlled.current === true && controlledState === undefined) {
+    if (controlled.current === true && !controlledValue) {
       warning(
         false,
-        'A component is changing a controlled component to be uncontrolled. ' +
+        'A component is changing a controlled %s component to be uncontrolled. ' +
```

Comment:
what does `%s` mean again?

Reply:
`%s` is a convention used in `warning`, it comes from `sprintf` style formatters in languages like C. You can also do it in things like `console.log()`, for example:

```js
console.log('Hello, %s', 'world'); // 'Hello, world'
```

Basically it's a way of interpolating values into a string